tbot_contrib/locking.py

```diff
@@ -0,0 +1,166 @@
+import contextlib
+import typing
+import tbot
+import time
+from tbot.machine import machine, linux
+
+
+class MachineLock(machine.PreConnectInitializer):
+    """
+    Machine locking implementation based on Python and bash.
+    """
+    @property
+    def lock_name(self) -> str:
+        return self.__class__.__name__
+
+    @property
+    def lock_expiry(self) -> int:
+        return -1
+
+    @property
+    def lock_pool_max(self) -> int:
+        return 1
+
+    @property
+    def lock_sleep(self) -> int:
+        return 1
+
+    lock_no = -1
+    lock_file = ""
+
+    @contextlib.contextmanager 
+    def _init_pre_connect(self) -> typing.Iterator:
+        with tbot.ctx.request(tbot.role.LabHost) as lh:
+            result = lh.exec("mktemp", "lock.XXX")
+            if result[0] == 0:
+                tempfile = result[1].strip()
```

Comment:
You probably want to make this a `Path`, then the `RedirStdout` below will also work:

```python
tempfile = linux.Path(lh, result[1].strip())
```